### PR TITLE
Identity columns

### PR DESCRIPTION
**Pull request by ericharmeling**
This PR documents new support for [PostgreSQL-like identity columns](https://www.postgresql.org/docs/current/sql-createtable.html). The PR also moves the "Column qualification" section from the `ADD COLUMN` page to the `CREATE TABLE` page, as more column qualifications are fully-supported in `CREATE TABLE` statements than in `ADD COLUMN` statements (e.g., Identity columns). 

Fixes https://github.com/cockroachdb/docs/issues/11417.
Fixes https://github.com/cockroachdb/docs/issues/11451.

### CODE REVIEW

**Review by ZhouXing19 — CHANGES_REQUESTED** [review]
Thanks for making this doc! Looks great, just added some nits.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/12189#-:-Mo45xbEERIQr080whNF:bbgi088)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ericharmeling and @rafiss)

___
*[v21.2/create-table.md, line 102 at r2](https://reviewable.io/reviews/cockroachdb/docs/12189#-Mo3oGN6-t-BhpWAmLoU:-Mo3oGN6-t-BhpWAmLoV:b74cj4) ([raw file](https://github.com/cockroachdb/docs/blob/e4531a5ca741e3487e5d856ffbe353233623ccb2/v21.2/create-table.md#L102)):*
> ```Markdown
> `IF NOT EXISTS` | Create a new table only if a table of the same name does not already exist in the database; if one does exist, do not return an error.<br><br>Note that `IF NOT EXISTS` checks the table name only; it does not check if an existing table has the same columns, indexes, constraints, etc., of the new table.
> `table_name` | The name of the table to create, which must be unique within its database and follow these [identifier rules](keywords-and-identifiers.html#identifiers). When the parent database is not set as the default, the name must be formatted as `database.name`.<br><br>The [`UPSERT`](upsert.html) and [`INSERT ON CONFLICT`](insert.html) statements use a temporary table called `excluded` to handle uniqueness conflicts during execution. It's therefore not recommended to use the name `excluded` for any of your tables.
> `column_def` | A comma-separated list of column definitions. Each column requires a [name/identifier](keywords-and-identifiers.html#identifiers) and [data type](data-types.html). Column names must be unique within the table but can have the same name as indexes or constraints.<br><br>You can optionally specify a [column qualification](#column-qualifications) (e.g., a [column-level constraint](constraints.html)). Any `PRIMARY KEY`, `UNIQUE`, and `CHECK` [constraints](constraints.html) defined at the column level are moved to the table-level as part of the table's creation. Use the [`SHOW CREATE`](show-create.html) statement to view them at the table level.
> ```

Looks like it missed `create-table.html` for `[column qualification]`.
___
*[v21.2/create-table.md, line 123 at r2](https://reviewable.io/reviews/cockroachdb/docs/12189#-Mo4-BSS0bRbSoKlULCX:-Mo4-BSS0bRbSoKlULCY:b-oim2dt) ([raw file](https://github.com/cockroachdb/docs/blob/e4531a5ca741e3487e5d856ffbe353233623ccb2/v21.2/create-table.md#L123)):*
> ```Markdown
> - [`DEFAULT` expressions](default-value.html)
> - <span class="version-tag">New in v21.2</span>: [`ON UPDATE` expressions](#on-update-expressions)
> - <span class="version-tag">New in v21.2</span>: [Identity columns](#identity-columns) ([sequence](create-sequence.html) columns)
> ```

nit: It might be better to use `[squence options]` than `[sequence]`.
___
*[v21.2/create-table.md, line 153 at r2](https://reviewable.io/reviews/cockroachdb/docs/12189#-Mo41CRf8mak_PWIfJia:-Mo41CRf8mak_PWIfJib:b-kangdz) ([raw file](https://github.com/cockroachdb/docs/blob/e4531a5ca741e3487e5d856ffbe353233623ccb2/v21.2/create-table.md#L153)):*
> ```Markdown
> If you use `GENERATED BY DEFAULT AS IDENTITY` to define the identity column, any [`INSERT`](insert.html)/[`UPSERT`](upsert.html)/[`UPDATE`](update.html) operations that specify a new value for the identity column will overwrite the default sequential values in the column. If you use `GENERATED ALWAYS AS IDENTITY`, the column's sequential values cannot be overwritten.
> 
> Note the following limitation of identity columns:
> ```

Should we also note that CockroachDB temporarily does not support the `OVERRIDING SYSTEM VALUE` syntax, which in PostgreSQL can be used in `INSERT/UPDATE/UPSERT` statements to overwrite a `GENERATED ALWAYS AS IDENTITY`? 
___
*[v21.2/create-table.md, line 942 at r2](https://reviewable.io/reviews/cockroachdb/docs/12189#-Mo43rsI4jzd8EW3ac_9:-Mo43rsI4jzd8EW3ac_A:b-q6lu8z) ([raw file](https://github.com/cockroachdb/docs/blob/e4531a5ca741e3487e5d856ffbe353233623ccb2/v21.2/create-table.md#L942)):*
The `numerical` column seems not to follow the insert order. Can we add an `order_index` INT8 column, and print the table ordered by it?

```
> CREATE TABLE bank (
                        ->     id UUID DEFAULT gen_random_uuid() PRIMARY KEY,
                        ->     order_index INT8 UNIQUE,
                        ->     balance INT8,
                        ->     payload STRING,
                        ->     numerical INT8 GENERATED BY DEFAULT AS IDENTITY (INCREMENT 1 MINVALUE 0 START 0)
                        -> );
~~~

> INSERT INTO bank (order_index, balance) VALUES (1, 0), (2, 0), (3, 0);
~~~

> SELECT id, order_index, balance, numerical FROM bank ORDER BY order_index;

~~~
                   id                  | order_index | balance | numerical
---------------------------------------+-------------+---------+------------
  c3a3b844-012b-4182-8a64-119abcea40b1 |           1 |       0 |         0
  1ce75589-78ef-46ef-8084-95b0e62dc11c |           2 |       0 |         1
  0f97ec06-480c-4705-9bf3-2eba6eaddb68 |           3 |       0 |         2

```
___
*[v21.2/create-table.md, line 966 at r2](https://reviewable.io/reviews/cockroachdb/docs/12189#-Mo45-Ri9l-yX5kdtdbS:-Mo45-Ri9l-yX5kdtdbT:b9a93fv) ([raw file](https://github.com/cockroachdb/docs/blob/e4531a5ca741e3487e5d856ffbe353233623ccb2/v21.2/create-table.md#L966)):*
Maybe we can add an example for `INSERT` too, to show that the inserted value won't affect the generation of the next sequential value: (I am following the table with `order_index`)

```
> INSERT INTO bank (order_index, balance, numerical) VALUES (4, 0, 3);
~~~

> INSERT INTO bank (order_index, balance) VALUES (5, 0);
~~~

> SELECT id, order_index, balance, numerical FROM bank ORDER BY order_index;
~~~

~~~
                   id                  | order_index | balance | numerical
---------------------------------------+-------------+---------+------------
  c3a3b844-012b-4182-8a64-119abcea40b1 |           1 |       0 |         0
  1ce75589-78ef-46ef-8084-95b0e62dc11c |           2 |       0 |         1
  0f97ec06-480c-4705-9bf3-2eba6eaddb68 |           3 |       0 |         2
  622d9fc4-0113-4ed3-bce9-8ace439cde08 |           4 |       0 |         3
  bb5a1363-4050-4d1e-8850-7ca3f0a6b3ae |           5 |       0 |         3
```

**Review by rafiss — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/12189)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ericharmeling, @rafiss, and @ZhouXing19)

___
*[v21.2/create-table.md, line 130 at r2](https://reviewable.io/reviews/cockroachdb/docs/12189#-Mo4lrUqADckqK8fchOD:-Mo4lrUqADckqK8fchOE:b-9m2edu) ([raw file](https://github.com/cockroachdb/docs/blob/e4531a5ca741e3487e5d856ffbe353233623ccb2/v21.2/create-table.md#L130)):*
> ```Markdown
> 
> - An [`UPDATE`](update.html) or [`UPSERT`](upsert.html) statement modifies a different column value in the same row.
> - An `ON UPDATE CASCADE` expression on a different column modifies an existing value in the same row.
> ```

could we make it more clear that "`ON UPDATE CASCADE` expression" is referring to a FK constraint definition -- and it is not referring to this new `ON UPDATE` expression itself

**Review by ericharmeling — COMMENTED** [review]
TFTR!

I made some updates in response to the great feedback.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/12189#-:-Mo580Nd3-GgE6HMS0YZ:bv390gw)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @ZhouXing19)

___
*[v21.2/create-table.md, line 102 at r2](https://reviewable.io/reviews/cockroachdb/docs/12189#-Mo3oGN6-t-BhpWAmLoU:-Mo4cC5V63yCWiQQmDVo:bms8c8q) ([raw file](https://github.com/cockroachdb/docs/blob/e4531a5ca741e3487e5d856ffbe353233623ccb2/v21.2/create-table.md#L102)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

Looks like it missed `create-table.html` for `[column qualification]`.
</blockquote></details>

(I think you're talking about `[column qualification](#column-qualifications)`?) This link is relative to the current page and so doesn't need `create-table.html` in the ref.

I know that's kind of confusing bc the file name is technically `create-table.md`, but when it's generated as HTML by the jekyll build, it becomes `create-table.html`. There are other examples on this page too (e.g., `For examples, see [Create a table with hash-sharded indexes](#create-a-table-with-a-hash-sharded-primary-index)`), and all over our docset.

Sorry if you knew that and are referring something else though! If so, LMK.
___
*[v21.2/create-table.md, line 130 at r2](https://reviewable.io/reviews/cockroachdb/docs/12189#-Mo4lrUqADckqK8fchOD:-Mo4v9dt14B74XTLvm1n:byvyago) ([raw file](https://github.com/cockroachdb/docs/blob/e4531a5ca741e3487e5d856ffbe353233623ccb2/v21.2/create-table.md#L130)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

could we make it more clear that "`ON UPDATE CASCADE` expression" is referring to a FK constraint definition -- and it is not referring to this new `ON UPDATE` expression itself
</blockquote></details>

Yes. Done.
___
*[v21.2/create-table.md, line 153 at r2](https://reviewable.io/reviews/cockroachdb/docs/12189#-Mo41CRf8mak_PWIfJia:-Mo4d2j35Jl7RAezPOZ_:b-4k0685) ([raw file](https://github.com/cockroachdb/docs/blob/e4531a5ca741e3487e5d856ffbe353233623ccb2/v21.2/create-table.md#L153)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

Should we also note that CockroachDB temporarily does not support the `OVERRIDING SYSTEM VALUE` syntax, which in PostgreSQL can be used in `INSERT/UPDATE/UPSERT` statements to overwrite a `GENERATED ALWAYS AS IDENTITY`? 
</blockquote></details>

Good point! Done.
___
*[v21.2/create-table.md, line 942 at r2](https://reviewable.io/reviews/cockroachdb/docs/12189#-Mo43rsI4jzd8EW3ac_9:-Mo4fv7R-kzpcjSJrh5d:b-44fpqm) ([raw file](https://github.com/cockroachdb/docs/blob/e4531a5ca741e3487e5d856ffbe353233623ccb2/v21.2/create-table.md#L942)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

The `numerical` column seems not to follow the insert order. Can we add an `order_index` INT8 column, and print the table ordered by it?

```
> CREATE TABLE bank (
                        ->     id UUID DEFAULT gen_random_uuid() PRIMARY KEY,
                        ->     order_index INT8 UNIQUE,
                        ->     balance INT8,
                        ->     payload STRING,
                        ->     numerical INT8 GENERATED BY DEFAULT AS IDENTITY (INCREMENT 1 MINVALUE 0 START 0)
                        -> );
~~~

> INSERT INTO bank (order_index, balance) VALUES (1, 0), (2, 0), (3, 0);
~~~

> SELECT id, order_index, balance, numerical FROM bank ORDER BY order_index;

~~~
                   id                  | order_index | balance | numerical
---------------------------------------+-------------+---------+------------
  c3a3b844-012b-4182-8a64-119abcea40b1 |           1 |       0 |         0
  1ce75589-78ef-46ef-8084-95b0e62dc11c |           2 |       0 |         1
  0f97ec06-480c-4705-9bf3-2eba6eaddb68 |           3 |       0 |         2

```
</blockquote></details>

Yeah that looks a little cleaner. Done.
___
*[v21.2/create-table.md, line 966 at r2](https://reviewable.io/reviews/cockroachdb/docs/12189#-Mo45-Ri9l-yX5kdtdbS:-Mo4fy6WD3LXKbYOA3YX:bhbobe4) ([raw file](https://github.com/cockroachdb/docs/blob/e4531a5ca741e3487e5d856ffbe353233623ccb2/v21.2/create-table.md#L966)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

Maybe we can add an example for `INSERT` too, to show that the inserted value won't affect the generation of the next sequential value: (I am following the table with `order_index`)

```
> INSERT INTO bank (order_index, balance, numerical) VALUES (4, 0, 3);
~~~

> INSERT INTO bank (order_index, balance) VALUES (5, 0);
~~~

> SELECT id, order_index, balance, numerical FROM bank ORDER BY order_index;
~~~

~~~
                   id                  | order_index | balance | numerical
---------------------------------------+-------------+---------+------------
  c3a3b844-012b-4182-8a64-119abcea40b1 |           1 |       0 |         0
  1ce75589-78ef-46ef-8084-95b0e62dc11c |           2 |       0 |         1
  0f97ec06-480c-4705-9bf3-2eba6eaddb68 |           3 |       0 |         2
  622d9fc4-0113-4ed3-bce9-8ace439cde08 |           4 |       0 |         3
  bb5a1363-4050-4d1e-8850-7ca3f0a6b3ae |           5 |       0 |         3
```
</blockquote></details>

Great idea. Done.

This behavior exposes another detail about identity columns that I neglected to document (i.e., that identity column creation creates a new sequence). I've updated this example, and expanded the description in the dedicated "Identity columns" section to account for it. PTAL

**Review by ZhouXing19 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/12189)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ericharmeling, @rafiss, and @ZhouXing19)

___
*[v21.2/create-table.md, line 102 at r2](https://reviewable.io/reviews/cockroachdb/docs/12189#-Mo3oGN6-t-BhpWAmLoU:-Mo59Ob32aI9sIx16KB7:b-rsjcmr) ([raw file](https://github.com/cockroachdb/docs/blob/e4531a5ca741e3487e5d856ffbe353233623ccb2/v21.2/create-table.md#L102)):*
<details><summary><i>Previously, ericharmeling (Eric Harmeling) wrote…</i></summary><blockquote>

(I think you're talking about `[column qualification](#column-qualifications)`?) This link is relative to the current page and so doesn't need `create-table.html` in the ref.

I know that's kind of confusing bc the file name is technically `create-table.md`, but when it's generated as HTML by the jekyll build, it becomes `create-table.html`. There are other examples on this page too (e.g., `For examples, see [Create a table with hash-sharded indexes](#create-a-table-with-a-hash-sharded-primary-index)`), and all over our docset.

Sorry if you knew that and are referring something else though! If so, LMK.
</blockquote></details>

Ah gotcha, thanks for the explanation!

**Review by ZhouXing19 — COMMENTED** [review]
LGTM, thanks!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/12189#-:-Mo59_OH6ZCFx4Ep5Qst:b2c0zcb)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ericharmeling, @rafiss, and @ZhouXing19)

File: v21.2/create-table.md [hunk 1]
```diff
@@ -111,6 +111,51 @@ Parameter | Description
...
+
+If you use `GENERATED BY DEFAULT AS IDENTITY` to define the identity column, any [`INSERT`](insert.html)/[`UPSERT`](upsert.html)/[`UPDATE`](update.html) operations that specify a new value for the identity column will overwrite the default sequential values in the column. If you use `GENERATED ALWAYS AS IDENTITY`, the column's sequential values cannot be overwritten.
+
+Note the following limitations of identity columns:
+
+- `GENERATED ALWAYS AS IDENTITY`/`GENERATED BY DEFAULT AS IDENTITY` is only supported in [`ALTER TABLE ... ADD COLUMN`](add-column.html) statements when the table being altered is empty, as [CockroachDB does not support back-filling sequential column data](known-limitations.html#adding-a-column-with-sequence-based-default-values). For more information, see the [Tracking GitHub Issue](https://github.com/cockroachdb/cockroach/issues/42508).
```
Comment (stbof) [inline v21.2/create-table.md:155]:
```suggestion
- `GENERATED ALWAYS AS IDENTITY`/`GENERATED BY DEFAULT AS IDENTITY` is supported in [`ALTER TABLE ... ADD COLUMN`](add-column.html) statements only when the table being altered is empty, as [CockroachDB does not support back-filling sequential column data](known-limitations.html#adding-a-column-with-sequence-based-default-values). For more information, see the [Tracking GitHub Issue](https://github.com/cockroachdb/cockroach/issues/42508).
```

File: v21.2/create-table.md [hunk 2]
```diff
@@ -111,6 +111,51 @@ Parameter | Description
...
+- [`DEFAULT` expressions](default-value.html)
+- <span class="version-tag">New in v21.2</span>: [`ON UPDATE` expressions](#on-update-expressions)
+- <span class="version-tag">New in v21.2</span>: [Identity columns](#identity-columns) (sequence-populated columns)
+
+### `ON UPDATE` expressions
+
+<span class="version-tag">New in v21.2</span>: `ON UPDATE` expressions update column values in the following cases:
+
+- An [`UPDATE`](update.html) or [`UPSERT`](upsert.html) statement modifies a different column value in the same row.
+- An `ON UPDATE CASCADE` [foreign key action](foreign-key.html#foreign-key-actions) on a different column modifies an existing value in the same row.
```
Comment (stbof) [inline v21.2/create-table.md:130]: The different structure of these two bullets make it harder to understand the behavior in the second case. Could the second one be rewritten as follows:

An `ON UPDATE CASCADE` [foreign key action](foreign-key.html#foreign-key-actions) modifies a different column value in the same row.

**Review by stbof — CHANGES_REQUESTED** [review]
See suggestion and comment/question.

**Review by rafiss — APPROVED** [review]
looks good from my end!

> Reviewed 1 of 2 files at r1, 3 of 4 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/12189#-:-Mo9iPWlEOEal5s_8XC1:b-5gbfxm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ericharmeling, @rafiss, and @ZhouXing19)

**Comment by stbof** [comment]
Should this feature be added to https://www.cockroachlabs.com/docs/v21.2/sql-feature-support.html#miscellaneous?

File: v21.2/create-table.md (hunk 1 above)
Comment (ericharmeling) [reply v21.2/create-table.md:155]: Done.

File: v21.2/create-table.md (hunk 2 above)
Comment (ericharmeling) [reply v21.2/create-table.md:130]: Done.

**Review by ericharmeling — COMMENTED** [review]
TFTR @stbof !

**Comment by ericharmeling** [comment]
> Should this feature be added to https://www.cockroachlabs.com/docs/v21.2/sql-feature-support.html#miscellaneous?

Probably. Good point. I think I'm going to open a separate PR for all of the feature support additions though, since there will be more on my end than just this one.

**Review by stbof — APPROVED** [review]

**Pull request merged**